Pinta.Gui.Widgets/Widgets/ColorPickerDialog.cs

```diff
@@ -0,0 +1,998 @@
+		sat_cps.SetValue (hsv.Sat * 100.0);
+		val_cps.SetValue (hsv.Val * 100.0);
+
+		r_cps.SetValue (CurrentColor.R * 255.0);
+		g_cps.SetValue (CurrentColor.G * 255.0);
+		b_cps.SetValue (CurrentColor.B * 255.0);
+		a_cps.SetValue (CurrentColor.A * 255.0);
+
+
+		// Update hex
+		if (GetFocus ()?.Parent != hex_entry)
+			hex_entry.SetText (CurrentColor.ToHex (hex_entry_add_alpha.state));
+
+		// Redraw palette displays
+		foreach (var display in color_displays) {
+			display.QueueDraw ();
+		}
+	}
+
+
+
+	private void DrawPaletteDisplay (Context g, Color c)
+	{
+		int xy = palette_display_border_thickness;
+		int wh = palette_display_size - palette_display_border_thickness * 2;
+		g.Antialias = Antialias.None;
+
+		// make checker pattern
+		if (c.A != 1) {
+			g.FillRectangle (new RectangleD (xy, xy, wh, wh), new Color (1, 1, 1));
+			g.FillRectangle (new RectangleD (xy, xy, wh / 2, wh / 2), new Color (.8, .8, .8));
+			g.FillRectangle (new RectangleD (xy + wh / 2, xy + wh / 2, wh / 2, wh / 2), new Color (.8, .8, .8));
+		}
+
+		g.FillRectangle (new RectangleD (xy, xy, wh, wh), c);
+		g.DrawRectangle (new RectangleD (xy, xy, wh, wh), new Color (0, 0, 0), palette_display_border_thickness);
+	}
+
+	private void DrawColorSurface (Context g)
+	{
+		int rad = picker_surface_radius;
+		int x0 = picker_surface_padding;
+		int y0 = picker_surface_padding;
+		int draw_width = picker_surface_radius * 2;
+		int draw_height = picker_surface_radius * 2;
+		int x1 = x0 + picker_surface_padding;
+		int y1 = y0 + picker_surface_padding;
+		PointI center = new PointI (rad, rad);
+
+		if (picker_surface_type == ColorSurfaceType.HueAndSat) {
+			int stride = draw_width * 4;
+
+			Span<byte> data = stackalloc byte[draw_height * stride];
+
+			for (int y = 0; y < draw_height; y++) {
+				for (int x = 0; x < draw_width; x++) {
+					PointI pxl = new PointI (x, y);
+					PointI vec = pxl - center;
+					if (vec.Magnitude () <= rad) {
+						var h = (MathF.Atan2 (vec.Y, -vec.X) + MathF.PI) / (2f * MathF.PI) * 360f;
+
+						var s = Math.Min (vec.Magnitude () / rad, 1);
+
+						double v = 1;
+						if (picker_surface_option_draw_value.state)
+							v = CurrentColor.ToHsv ().Val;
+
+						var c = Color.FromHsv (h, s, v);
+
+						double a = 1;
+						var d = rad - vec.Magnitude ();
+						if (d < 1) {
+							a = d;
+						}
+
+						data[(y * stride) + (x * 4) + 0] = (byte) (c.R * 255);
+						data[(y * stride) + (x * 4) + 1] = (byte) (c.G * 255);
+						data[(y * stride) + (x * 4) + 2] = (byte) (c.B * 255);
+						data[(y * stride) + (x * 4) + 3] = (byte) (a * 255);
+					} else {
+						data[(y * stride) + (x * 4) + 0] = (byte) (0);
+						data[(y * stride) + (x * 4) + 1] = (byte) (0);
+						data[(y * stride) + (x * 4) + 2] = (byte) (0);
+						data[(y * stride) + (x * 4) + 3] = (byte) (0);
+					}
+				}
+			}
+
+			var img = MemoryTexture.New (draw_width, draw_height, MemoryFormat.R8g8b8a8, Bytes.New (data), (UIntPtr) stride).ToSurface ();
```

Comment:
There are a couple extra copies of the buffer here, so it would probably be more straightforward to create an `ImageSurface` of the right dimensions, and then use `GetPixelData()` to write directly to its pixels